### PR TITLE
Add remote shards allocator for searchable snapshots

### PR DESCRIPTION
**Pull request by kotwanikunal**
Signed-off-by: Kunal Kotwani <kkotwani@amazon.com>

### Description
- Adds a remote allocator for searchable snapshots

### Issues Resolved
- #4777 

### Check List
- [x] New functionality includes testing.
  - [x] All tests pass
- [x] New functionality has been documented.
  - [x] New functionality has javadoc added
- [x] Commits are signed per the DCO using --signoff
- [x] Commit changes are listed out in CHANGELOG.md file (See: [Changelog](../blob/main/CONTRIBUTING.md#changelog))

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: server/src/main/java/org/opensearch/common/util/FeatureFlags.java [hunk 1]
```diff
@@ -43,6 +43,9 @@ public class FeatureFlags {
      * and false otherwise.
      */
     public static boolean isEnabled(String featureFlagName) {
+        if (SEARCHABLE_SNAPSHOT.equals(featureFlagName)) {
```
Comment (kotwanikunal) [inline server/src/main/java/org/opensearch/common/util/FeatureFlags.java:46]: I need to figure out a way to enable this flag for tests and remove this shim.

File: server/src/main/java/org/opensearch/cluster/routing/RoutingPool.java [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+
+package org.opensearch.cluster.routing;
+
+import org.opensearch.cluster.metadata.IndexMetadata;
+import org.opensearch.cluster.node.DiscoveryNode;
+import org.opensearch.cluster.routing.allocation.RoutingAllocation;
+import org.opensearch.common.settings.Settings;
+import org.opensearch.index.IndexModule;
+
+public enum RoutingPool {
```
Comment (andrross) [inline server/src/main/java/org/opensearch/cluster/routing/RoutingPool.java:17]: Please add documentation to all new classes. Be sure to include the internal tag as well.

Comment (kotwanikunal) [reply server/src/main/java/org/opensearch/cluster/routing/RoutingPool.java:24]: Done.

**Comment by kotwanikunal** [comment]
Unrelated failures: 
```
> [2022-10-21T23:41:37,721][INFO ][o.o.b.MixedClusterClientYamlTestSuiteIT] [test] [p0=index/80_geo_point/Single point test] after test
  2> REPRODUCE WITH: ./gradlew ':qa:mixed-cluster:v2.4.0#mixedClusterTest' --tests "org.opensearch.backwards.MixedClusterClientYamlTestSuiteIT.test {p0=index/80_geo_point/Single point test}" -Dtests.seed=5BA55F046123AF6F -Dtests.security.manager=true -Dtests.jvm.argline="-XX:TieredStopAtLevel=1 -XX:ReservedCodeCacheSize=64m" -Dtests.locale=fr-CH -Dtests.timezone=Antarctica/Syowa -Druntime.java=17
  2> java.lang.AssertionError: Failure at [index/80_geo_point:60]: hits.total didn't match expected value:
```

```
Tests with failures:
 - org.opensearch.backwards.MixedClusterClientYamlTestSuiteIT.test {p0=index/80_geo_point/Multi points test}
 - org.opensearch.backwards.MixedClusterClientYamlTestSuiteIT.test {p0=index/80_geo_point/Single point test}
 ```

**Comment by kotwanikunal** [comment]
```
Tests with failures:
 - org.opensearch.backwards.MixedClusterClientYamlTestSuiteIT.test {p0=index/80_geo_point/Single point test}
 - org.opensearch.backwards.MixedClusterClientYamlTestSuiteIT.test {p0=index/80_geo_point/Multi points test}
```

**Comment by kotwanikunal** [comment]
#4852

File: server/src/main/java/org/opensearch/cluster/routing/allocation/allocator/RemoteShardsBalancer.java [hunk 3]
```diff
@@ -0,0 +1,578 @@
...
+                    allocation.changes()
+                );
+            }
+        }
+    }
+
+    private Queue<RoutingNode> getShuffledRemoteNodes() {
+        Iterator<RoutingNode> nodesIter = routingNodes.mutableIterator();
+        List<RoutingNode> nodeList = new ArrayList<>();
+        while (nodesIter.hasNext()) {
```
Comment (andrross) [inline server/src/main/java/org/opensearch/cluster/routing/allocation/allocator/RemoteShardsBalancer.java:545]: Can you replace the mutable iterator and while loop with the following?

```
for (RoutingNode rNode : routingNodes) {
    ...
}
```

File: server/src/main/java/org/opensearch/cluster/routing/allocation/allocator/RemoteShardsBalancer.java [hunk 4]
```diff
@@ -0,0 +1,578 @@
...
+                            break;
+                        }
+                    }
+                }
+            }
+        }
+    }
+
+    /**
+     * Performs heuristic based balancing for remote shards within the cluster.
```
Comment (andrross) [inline server/src/main/java/org/opensearch/cluster/routing/allocation/allocator/RemoteShardsBalancer.java:180]: Can you expand a bit here? For example, what is the heuristic being used here? Doesn't need to be super detailed, just a high level description of exactly what is being done.

File: server/src/main/java/org/opensearch/cluster/routing/allocation/allocator/RemoteShardsBalancer.java [hunk 5]
```diff
@@ -0,0 +1,578 @@
...
+            String index = shard.getIndexName();
+            if (!RoutingPool.REMOTE_CAPABLE.equals(RoutingPool.getShardPool(shard, allocation))) {
+                routingNodes.unassigned().add(shard);
+                continue;
+            }
+            if (!unassignedShardMap.containsKey(index)) {
+                unassignedShardMap.put(index, new UnassignedIndexShards());
+            }
+            UnassignedIndexShards indexShards = unassignedShardMap.get(index);
+            indexShards.addShard(shard);
```
Comment (andrross) [inline server/src/main/java/org/opensearch/cluster/routing/allocation/allocator/RemoteShardsBalancer.java:259]: Nitpick, but can probably just do:

```
unassignedShardMap.get(index).addShard(shard);
```

**Comment by kotwanikunal** [comment]
Seems related since the flag is forcefully turned on for `SEARCHABLE_SNAPSHOT` -
```
Tests with failures:
 - org.opensearch.backwards.MixedClusterClientYamlTestSuiteIT.test {p0=snapshot.restore/10_basic/Create a snapshot and then restore it}

```

File: server/src/main/java/org/opensearch/cluster/routing/allocation/allocator/RemoteShardsBalancer.java (hunk 4 above)
Comment (kotwanikunal) [reply server/src/main/java/org/opensearch/cluster/routing/allocation/allocator/RemoteShardsBalancer.java]: Done!

File: server/src/main/java/org/opensearch/common/util/FeatureFlags.java (hunk 1 above)
Comment (kotwanikunal) [reply server/src/main/java/org/opensearch/common/util/FeatureFlags.java]: Updated to use `System` method.

**Comment by kotwanikunal** [comment]
Unrelated failure:
```
 - org.opensearch.smoketest.SmokeTestMultiNodeClientYamlTestSuiteIT.test {yaml=search/30_limits/Regexp length limit}
```

**Comment by kotwanikunal** [comment]
Hmm. Same test failure
```
REPRODUCE WITH: ./gradlew ':qa:smoke-test-multinode:integTest' --tests "org.opensearch.smoketest.SmokeTestMultiNodeClientYamlTestSuiteIT.test {yaml=search/30_limits/Regexp length limit}" -Dtests.seed=45E07907A235AA6A -Dtests.security.manager=true -Dtests.jvm.argline="-XX:TieredStopAtLevel=1 -XX:ReservedCodeCacheSize=64m" -Dtests.locale=es-HN -Dtests.timezone=SystemV/PST8PDT -Druntime.java=17

org.opensearch.smoketest.SmokeTestMultiNodeClientYamlTestSuiteIT > test {yaml=search/30_limits/Regexp length limit} FAILED
    java.lang.AssertionError: Failure at [search/30_limits:121]: the error message was expected to match the provided regex but didn't
```

**Comment by kotwanikunal** [comment]
#4899

File: CHANGELOG.md [hunk 6]
```diff
@@ -69,10 +69,14 @@ Inspired from [Keep a Changelog](https://keepachangelog.com/en/1.0.0/)
 - Update Jackson Databind to 2.13.4.2 (addressing CVE-2022-42003) ([#4779](https://github.com/opensearch-project/OpenSearch/pull/4779))
 - Bumps `tika` from 2.4.0 to 2.5.0 ([#4791](https://github.com/opensearch-project/OpenSearch/pull/4791))
 - Exclude jettison version brought in with hadoop-minicluster. ([#4787](https://github.com/opensearch-project/OpenSearch/pull/4787))
+<<<<<<< HEAD
```
Comment (andrross) [inline CHANGELOG.md:72]: ugh

**Comment by kotwanikunal** [comment]
Failed with - 
```
Tests with failures:
 - org.opensearch.repositories.s3.RepositoryS3ClientYamlTestSuiteIT.test {yaml=repository_s3/20_repository_permanent_credentials/Get a non existing snapshot}
 ```

**Comment by kotwanikunal** [comment]
> ### Gradle Check (Jenkins) Run Completed with:
> * **RESULT:** UNSTABLE ❌
> * **URL:** https://build.ci.opensearch.org/job/gradle-check/5733/
> * **CommitID:** [ca76925](https://github.com/opensearch-project/OpenSearch/commit/ca76925215a8d1dadc3f98599f408defdeeca433)
>   Please examine the workflow log, locate, and copy-paste the failure below, then iterate to green.
>   Is the failure [a flaky test](https://github.com/opensearch-project/OpenSearch/blob/main/DEVELOPER_GUIDE.md#flaky-tests) unrelated to your change?

#3603

**Review by andrross — APPROVED** [review]

File: server/src/main/java/org/opensearch/cluster/routing/allocation/allocator/RemoteShardsBalancer.java [hunk 7]
```diff
@@ -0,0 +1,598 @@
...
+                throttledNodeCount
+            );
+        }
+
+        if (excludedNodes.isEmpty()) {
+            logger.debug("No excluded nodes found. Returning...");
+            return;
+        }
+
+        while (!eligibleNodes.isEmpty() && !excludedNodes.isEmpty()) {
```
Comment (Vishalks) [inline server/src/main/java/org/opensearch/cluster/routing/allocation/allocator/RemoteShardsBalancer.java:113]: Is there a way we could simplify this code? While under a for under a while looks to be of higher time complexity, although I am not sure if there's another way to simplify it.

File: server/src/main/java/org/opensearch/cluster/routing/allocation/allocator/RemoteShardsBalancer.java [hunk 8]
```diff
@@ -0,0 +1,598 @@
...
+        allocateUnassignedPrimaries(nodeQueue, unassignedShardMap);
+        allocateUnassignedReplicas(nodeQueue, unassignedShardMap);
+        ignoreRemainingShards(unassignedShardMap);
+    }
+
+    /**
+     * Performs shard movement for incompatible remote shards
+     */
+    @Override
+    void moveShards() {
```
Comment (Vishalks) [inline server/src/main/java/org/opensearch/cluster/routing/allocation/allocator/RemoteShardsBalancer.java:81]: Can we break this method to smaller ones for better readability?

File: server/src/main/java/org/opensearch/cluster/routing/allocation/allocator/RemoteShardsBalancer.java [hunk 9]
```diff
@@ -0,0 +1,598 @@
...
+
+        if (nodeList.isEmpty()) {
+            logger.info("No eligible remote nodes found to perform balancing");
+            return;
+        }
+
+        ObjectIntHashMap<String> primaryShardCount = new ObjectIntHashMap<>();
+        int totalPrimaryShard = 0;
+        for (RoutingNode node : nodeList) {
+            int totalPrimaryPerNode = 0;
```
Comment (Vishalks) [inline server/src/main/java/org/opensearch/cluster/routing/allocation/allocator/RemoteShardsBalancer.java:202]: What's the difference between totalPrimaryShard above and totalPrimaryPerNode? If the latter conveys totalPrimaryShardPerNode, we can rename it accordingly. If not, maybe a better variable name to distinguish would help here?

File: server/src/main/java/org/opensearch/cluster/routing/allocation/allocator/RemoteShardsBalancer.java [hunk 10]
```diff
@@ -0,0 +1,598 @@
...
+        logger.trace("Performing balancing for remote shards.");
+
+        if (nodeList.isEmpty()) {
+            logger.info("No eligible remote nodes found to perform balancing");
+            return;
+        }
+
+        ObjectIntHashMap<String> primaryShardCount = new ObjectIntHashMap<>();
+        int totalPrimaryShard = 0;
+        for (RoutingNode node : nodeList) {
```
Comment (Vishalks) [inline server/src/main/java/org/opensearch/cluster/routing/allocation/allocator/RemoteShardsBalancer.java:201]: Can we extract this logic to a different method?

File: server/src/main/java/org/opensearch/cluster/routing/allocation/allocator/RemoteShardsBalancer.java [hunk 11]
```diff
@@ -0,0 +1,598 @@
...
+            for (ShardRouting shard : indexShards.getPrimaries()) {
+                routingNodes.unassigned().ignoreShard(shard, UnassignedInfo.AllocationStatus.DECIDERS_NO, allocation.changes());
+            }
+            for (ShardRouting shard : indexShards.getReplicas()) {
+                routingNodes.unassigned().ignoreShard(shard, UnassignedInfo.AllocationStatus.DECIDERS_NO, allocation.changes());
+            }
+        }
+    }
+
+    private void allocateUnassigned(
```
Comment (Vishalks) [inline server/src/main/java/org/opensearch/cluster/routing/allocation/allocator/RemoteShardsBalancer.java:301]: Can we break this down to smaller methods? Looks difficult to follow.

Comment (kotwanikunal) [reply server/src/main/java/org/opensearch/cluster/routing/allocation/allocator/RemoteShardsBalancer.java]: Done.

File: server/src/main/java/org/opensearch/cluster/routing/allocation/allocator/RemoteShardsBalancer.java (hunk 10 above)
Comment (kotwanikunal) [reply server/src/main/java/org/opensearch/cluster/routing/allocation/allocator/RemoteShardsBalancer.java]: Done.

File: server/src/main/java/org/opensearch/cluster/routing/allocation/allocator/RemoteShardsBalancer.java (hunk 9 above)
Comment (kotwanikunal) [reply server/src/main/java/org/opensearch/cluster/routing/allocation/allocator/RemoteShardsBalancer.java]: Renamed. It was exactly what you figured out. 
Simplified the logic and extracted it out.

File: server/src/main/java/org/opensearch/cluster/routing/allocation/allocator/RemoteShardsBalancer.java (hunk 7 above)
Comment (kotwanikunal) [reply server/src/main/java/org/opensearch/cluster/routing/allocation/allocator/RemoteShardsBalancer.java]: This was pointed out as one of the misses within the design itself. Currently we are looking to re-use the correctness of the existing approach.
We will work on optimizing this as a part of phase 2. 
For more references: #4759

**Comment by kotwanikunal** [comment]
> ### Gradle Check (Jenkins) Run Completed with:
> * **RESULT:** FAILURE ❌
> * **URL:** https://build.ci.opensearch.org/job/gradle-check/5832/
> * **CommitID:** [ab5ce2e](https://github.com/opensearch-project/OpenSearch/commit/ab5ce2eae413cad4ef5a41d096465ceb8f0e81f9)
>   Please examine the workflow log, locate, and copy-paste the failure below, then iterate to green.
>   Is the failure [a flaky test](https://github.com/opensearch-project/OpenSearch/blob/main/DEVELOPER_GUIDE.md#flaky-tests) unrelated to your change?

Unrelated failure: 
```
REPRODUCE WITH: ./gradlew ':sandbox:plugins:concurrent-search:test' --tests "org.opensearch.search.query.QueryProfilePhaseTests" -Dtests.method="testTerminateAfterEarlyTermination {p0=5 p1=org.opensearch.search.query.ConcurrentQueryPhaseSearcher@1d6a22dd}" -Dtests.seed=E4EDD6626F1EC93B -Dtests.security.manager=true -Dtests.jvm.argline="-XX:TieredStopAtLevel=1 -XX:ReservedCodeCacheSize=64m" -Dtests.locale=hu-HU -Dtests.timezone=SystemV/PST8PDT -Druntime.java=17

org.opensearch.search.query.QueryProfilePhaseTests > testTerminateAfterEarlyTermination {p0=5 p1=org.opensearch.search.query.ConcurrentQueryPhaseSearcher@1d6a22dd} FAILED
    java.lang.AssertionError: 
    Expected: a value greater than <0L>
         but: <0L> was equal to <0L>
        at __randomizedtesting.SeedInfo.seed([E4EDD6626F1EC93B:72496FF5662CB552]:0)
        at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:18)
        at org.junit.Assert.assertThat(Assert.java:964)
        at org.junit.Assert.assertThat(Assert.java:930)
        at org.opensearch.search.query.QueryProfilePhaseTests.lambda$testTerminateAfterEarlyTermination$26(QueryProfilePhaseTests.java:574)
        at org.opensearch.search.query.QueryProfilePhaseTests.lambda$assertProfileData$55(QueryProfilePhaseTests.java:1105)
        at org.opensearch.search.query.QueryProfilePhaseTests.assertProfileData(QueryProfilePhaseTests.java:1156)
        at org.opensearch.search.query.QueryProfilePhaseTests.assertProfileData(QueryProfilePhaseTests.java:1111)
        at org.opensearch.search.query.QueryProfilePhaseTests.assertProfileData(QueryProfilePhaseTests.java:1102)
        at org.opensearch.search.query.QueryProfilePhaseTests.testTerminateAfterEarlyTermination(QueryProfilePhaseTests.java:555)

```

**Comment by kotwanikunal** [comment]
> 

#4881

**Pull request merged**